pkg/controller/ingress-shim/controller_test.go

```diff
@@ -0,0 +1,163 @@
+/*
+Copyright 2020 The cert-manager Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package controller
+
+import (
+	"context"
+	"testing"
+	"time"
+
+	testpkg "github.com/jetstack/cert-manager/pkg/controller/test"
+	"github.com/stretchr/testify/assert"
+	"github.com/stretchr/testify/require"
+	networkingv1beta1 "k8s.io/api/networking/v1beta1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/runtime"
+	kclient "k8s.io/client-go/kubernetes"
+
+	cmapi "github.com/jetstack/cert-manager/pkg/apis/certmanager/v1"
+	cmclient "github.com/jetstack/cert-manager/pkg/client/clientset/versioned"
+)
+
+var _ = cmapi.Certificate{}
+
+func Test_controller_Register(t *testing.T) {
+	tests := []struct {
+		name              string
+		existingKObjects  []runtime.Object
+		existingCMObjects []runtime.Object
+		givenCall         func(*testing.T, cmclient.Interface, kclient.Interface)
+		expectRequeueKey  string
+	}{
+		{
+			name: "ingress should be queued when it is created",
+			givenCall: func(t *testing.T, _ cmclient.Interface, c kclient.Interface) {
+				_, err := c.NetworkingV1beta1().Ingresses("namespace-1").Create(context.Background(), &networkingv1beta1.Ingress{ObjectMeta: metav1.ObjectMeta{
+					Namespace: "namespace-1", Name: "ingress-1",
+				}}, metav1.CreateOptions{})
+				require.NoError(t, err)
+			},
+			expectRequeueKey: "namespace-1/ingress-1",
+		},
+		{
+			name: "ingress should be queued when it is updated",
+			existingKObjects: []runtime.Object{&networkingv1beta1.Ingress{ObjectMeta: metav1.ObjectMeta{
+				Namespace: "namespace-1", Name: "ingress-1",
+			}}},
+			givenCall: func(t *testing.T, _ cmclient.Interface, c kclient.Interface) {
+				_, err := c.NetworkingV1beta1().Ingresses("namespace-1").Update(context.Background(), &networkingv1beta1.Ingress{ObjectMeta: metav1.ObjectMeta{
+					Namespace: "namespace-1", Name: "ingress-1",
+				}}, metav1.UpdateOptions{})
+				require.NoError(t, err)
+			},
+			expectRequeueKey: "namespace-1/ingress-1",
+		},
+		{
+			name: "ingress should be queued when it is deleted",
```

Comment:
Question: I keep forgetting why we need this. We would expect the certificate to be garbage collected when the ingress is deleted- why do we queue it at this point?

Reply:
You are correct, we don't need to requeue the Ingress when receiving the "Deleted" event.

Also I was wrong when I told you that we need to requeue on "Deleted" events when there is a finalizer. That's not true because marking an object for deletion does not emit "Deleted"; instead, it emits "Updated". "Deleted" is only emitted after the apiserver has removed the object entirely from etcd.

I think we don't need to requeue upon "Deleted".

Reply:
I added a comment that looks like this:
```go
	// We still requeue on "Deleted" for consistency with the rest of the
	// controllers, but we don't actually need to. "Deleted" is only emitted
	// after the apiserver has removed the object entirely from etcd; if we had
	// to do some cleanup, we would use a finalizer (which triggers a "Updated"
	// event).
	kShared.Networking().V1beta1().Ingresses().Informer().AddEventHandler(&controllerpkg.QueuingEventHandler{
		Queue: queue,
	})
```